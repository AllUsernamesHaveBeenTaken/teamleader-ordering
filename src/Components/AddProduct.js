import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddProduct extends Component {
  constructor(props) {
    super(props);
    const { products } = this.props;
    this.state = {
      quantity: 1,
      selectedItem: products[0],
    };
  }

  selectItem = (e) => {
    const { products } = this.props;
    this.setState({
      quantity: 1,
      selectedItem: products.find(item => item.id === e.target.value),
    });
  };

  render() {
    const { quantity, selectedItem } = this.state;
    const { products, addNewItem, orderId } = this.props;
    return (
      <div>
        <h3>Product</h3>
        <select onChange={this.selectItem}>
          {products.map(item => (
            <option key={`product-${item.id}`} value={item.id}>
              {item.id}
            </option>
          ))}
        </select>
        <h3>Description</h3>
        <p>{selectedItem.description}</p>
        <h3>Quantity</h3>
        <input
          type="number"
          placeholder={quantity}
          value={quantity}
          min="1"
          onChange={e => this.setState({ quantity: e.target.value < 1 ? 1 : e.target.value })}
        />
        <h3>{`Total: ${(+quantity * selectedItem.price).toFixed(2)}`}</h3>
        <button
          type="submit"
          onClick={() =>
            addNewItem(orderId, {
              'product-id': selectedItem.id,
              quantity: quantity.toString(),
              'unit-price': selectedItem.price,
              total: (+quantity * selectedItem.price).toFixed(2),
            })
          }
        >
          Add to cart
        </button>
      </div>
    );
  }
}

export default AddProduct;

AddProduct.propTypes = {
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      description: PropTypes.string,
      category: PropTypes.string,
      price: PropTypes.string,
    }),
  ).isRequired,
  orderId: PropTypes.string.isRequired,
  addNewItem: PropTypes.func.isRequired,
};
