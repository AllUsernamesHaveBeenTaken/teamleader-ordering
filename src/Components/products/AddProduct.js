import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';

class AddProduct extends PureComponent {
  constructor(props) {
    super(props);
    const { products } = this.props;
    this.state = {
      quantity: 1,
      selectedItem: products[0],
    };
  }

  selectItem = (e) => {
    const { products } = this.props;
    this.setState({
      quantity: 1,
      selectedItem: products.find(item => item.id === e.target.value),
    });
  };

  render() {
    const { quantity, selectedItem } = this.state;
    const { products, addNewItem, orderId } = this.props;
    return (
      <Fragment>
        <h3>Product</h3>
        <select onChange={this.selectItem} data-testid="select-product">
          {products.map(item => (
            <option key={`product-${item.id}`} value={item.id}>
              {item.description}
            </option>
          ))}
        </select>
        <h3>Quantity</h3>
        <input
          data-testid="quantity-product"
          type="number"
          placeholder={quantity}
          value={quantity}
          min="1"
          onChange={e => this.setState({ quantity: e.target.value < 1 ? 1 : e.target.value })}
        />
        <h3>{`Total: ${(+quantity * selectedItem.price).toFixed(2)}`}</h3>
        <button
          data-testid="submit-new-product"
          type="submit"
          onClick={() =>
            addNewItem(orderId, {
              'product-id': selectedItem.id,
              quantity: quantity.toString(),
              'unit-price': selectedItem.price,
              total: (+quantity * selectedItem.price).toFixed(2),
            })
          }
        >
          Add to cart
        </button>
      </Fragment>
    );
  }
}

export default AddProduct;

AddProduct.propTypes = {
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      description: PropTypes.string,
      category: PropTypes.string,
      price: PropTypes.string,
    }),
  ).isRequired,
  orderId: PropTypes.string.isRequired,
  addNewItem: PropTypes.func.isRequired,
};
