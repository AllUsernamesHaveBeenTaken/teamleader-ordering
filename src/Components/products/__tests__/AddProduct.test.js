import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import AddProduct from '../AddProduct';

afterEach(() => {
  cleanup();
  console.error.mockClear();
});

console.error = jest.fn();

const addNewItem = jest.fn();

const orderId = '1';

const products = [
  {
    id: 'A101',
    description: 'Screwdriver',
    category: '1',
    price: '9.75',
  },
  {
    id: 'A102',
    description: 'Electric screwdriver',
    category: '1',
    price: '49.50',
  },
];

// AddProduct's props = products, addNewItem, orderId
test('<AddProduct />', () => {
  const { debug, getByTestId } = render(
    <AddProduct products={products} addNewItem={addNewItem} orderId={orderId} />,
  );

  expect(getByTestId('select-product')).toBeTruthy();
  getByTestId('select-product').value = 'A101';
  fireEvent.change(getByTestId('select-product'));

  expect(getByTestId('quantity-product')).toBeTruthy();
  getByTestId('quantity-product').value = '1';
  fireEvent.change(getByTestId('quantity-product'));

  expect(getByTestId('submit-new-product')).toBeTruthy();
  fireEvent.click(getByTestId('submit-new-product'));
  expect(addNewItem).toHaveBeenCalledTimes(1);
  expect(addNewItem).toBeCalledWith(orderId, {
    'product-id': 'A101',
    quantity: '1',
    'unit-price': '9.75',
    total: '9.75',
  });

  debug();
});
