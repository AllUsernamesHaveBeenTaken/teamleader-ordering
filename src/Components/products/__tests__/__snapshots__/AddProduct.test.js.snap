// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddProduct /> 1`] = `
<h3>
  Product
</h3>
`;
